harvester/pcidevices
# Feat: add USB Passthrough

**Problem:**

This is a new feature request to make use of [KubeVirt USB Passthrough on the node](https://kubevirt.io/user-guide/virtual_machines/host-devices/#usb-host-passthrough).

**Solution:**

Expose KubeVirt config and create custom external device plugin to select specified usb device.

**Related Issue:**

https://github.com/harvester/harvester/issues/5762

**Guideline to this PR**

This PR is quite large, so I list some files as entry for reviewing.

- New APIs
  - pkg/apis/devices.harvesterhci.io/v1beta1/usb_device.go
  - pkg/apis/devices.harvesterhci.io/v1beta1/usb_device_claim.go
- Controllers
  - Could check test cases first.
  - pkg/controller/usbdevice/usbdevice_controller.go
  - pkg/controller/usbdevice/usbdevice_claim_controller.go
- Custom USB Device Plugin
  - pkg/deviceplugins/usb_device_plugin.go
- Retrive USB detail informations
  - pkg/util/gousb/usbid/load.go
- Webhook
  - pkg/webhook/usbdevice.go
  - pkg/webhook/usbdeviceclaim.go
  - pkg/webhook/vmi_validatation.go
  - pkg/webhook/pcideviceclaim.go

Other files are targeted to different purposes, such as rebooting,  code generator and fixing codegen.

This is a very simple arch of objects. 

![image](https://github.com/user-attachments/assets/912262a3-b153-48f7-8c47-49de2433886b)


**Test plan:**

Please find machine with USB device on the node to test this feature.

This is happy path: 
1.  There should be devices on the `usbdevice` custom resource.
2. Enable it by creating like following custom resource
```yaml
apiVersion: devices.harvesterhci.io/v1beta1
kind: USBDeviceClaim
metadata:
  name: {usbdevice.Name}
  ownerReferences:
  - apiVersion: devices.harvesterhci.io/v1beta1
    kind: USBDevice
    name: {usbdevice.Name}
    uid: {usbdevice.UID}
```
3. Create a VM, and stop it. Then modify its yaml.
```
domain:
  device:
    hostDevices:
      - deviceName: kubevirt.io/{usbdevice.Name}
         name: {usbdevice.Name}
```
4. Start vm, and run `lsusb` to check device.

The others paths, including:
1. Two same vendor and product, should be able pick one of them.
2. Delete the `usbdeviceclaim` which is being used, should not be deleted.
3. Delete the `usbdevice` which is being used, should not be deleted.
4. Enable `pcidevice` after its `usbdevice` is claimed, should not be enabled.
